cdc/sink/codec/encoder_group.go

```diff
@@ -0,0 +1,171 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codec
+
+import (
+	"context"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sink/codec/common"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	defaultEncoderGroupSize = 16
+	defaultInputChanSize    = 256
+	defaultMetricInterval   = 15 * time.Second
+)
+
+// EncoderGroup manages a group of encoders
+type EncoderGroup interface {
+	// Run start the group
+	Run(ctx context.Context) error
+	// AddEvent add an event into the group, handled by one of the encoders
+	AddEvent(ctx context.Context, topic string, partition int32, event *model.RowChangedEvent, callback func()) error
+	// Output returns a channel produce futures
+	Output() <-chan *future
+}
+
+type encoderGroup struct {
+	changefeedID model.ChangeFeedID
+
+	builder EncoderBuilder
+	count   int
+	inputCh []chan *future
+	index   uint64
+
+	outputCh chan *future
+}
+
+// NewEncoderGroup creates a new EncoderGroup instance
+func NewEncoderGroup(builder EncoderBuilder, count int, changefeedID model.ChangeFeedID) *encoderGroup {
+	if count <= 0 {
+		count = defaultEncoderGroupSize
+	}
+
+	inputCh := make([]chan *future, count)
+	for i := 0; i < count; i++ {
+		inputCh[i] = make(chan *future, defaultInputChanSize)
+	}
+
+	return &encoderGroup{
+		changefeedID: changefeedID,
+
+		builder:  builder,
+		count:    count,
+		inputCh:  inputCh,
+		index:    0,
+		outputCh: make(chan *future, defaultInputChanSize*count),
+	}
+}
+
+func (g *encoderGroup) Run(ctx context.Context) error {
+	defer func() {
+		encoderGroupInputChanSizeGauge.DeleteLabelValues(g.changefeedID.Namespace, g.changefeedID.ID)
+		close(g.outputCh)
+		log.Info("encoder group exited",
+			zap.String("namespace", g.changefeedID.Namespace),
+			zap.String("changefeed", g.changefeedID.ID))
+	}()
+	eg, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < g.count; i++ {
+		idx := i
+		eg.Go(func() error {
+			return g.runEncoder(ctx, idx)
+		})
+	}
+	return eg.Wait()
+}
+
+func (g *encoderGroup) runEncoder(ctx context.Context, idx int) error {
+	encoder := g.builder.Build()
+	inputCh := g.inputCh[idx]
+	metric := encoderGroupInputChanSizeGauge.
+		WithLabelValues(g.changefeedID.Namespace, g.changefeedID.ID, strconv.Itoa(idx))
+	ticker := time.NewTicker(defaultMetricInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			metric.Set(float64(len(inputCh)))
+		case future := <-inputCh:
+			if err := encoder.AppendRowChangedEvent(ctx, future.Topic, future.Event, future.Callback); err != nil {
+				return err
+			}
+			future.Messages = encoder.Build()
+			close(future.done)
+		}
+	}
+}
+
+func (g *encoderGroup) AddEvent(ctx context.Context, topic string, partition int32, event *model.RowChangedEvent, callback func()) error {
```

Comment:
I guess we can combine `topic string, partition int32, event *model.RowChangedEvent, callback func()`, because we copy it again and again.

Reply:
they were extracted from `mqEvent`. we cannot pass `mqEvent` directly since it introduces cyclic import problem.

If we move mqEvent to codec package, it becomes more complicated and weird

Reply:
In any case, it is not good to copy the same relevant parameters four times :(